models_evaluation.py

```diff
@@ -0,0 +1,176 @@
+from os import listdir
+from os.path import isfile, join
+import sys
+import argparse
+import logging
+import pathlib
+
+import gensim
+import pandas as pd
+import itertools
+
+logger = logging.getLogger("fasttext")
+
+
+def questions_reading(analogies):
+    """
+    Reads all type of questions to dict.
+    :param analogies: string
+    :return: dict
+    """
+    analogy_dict = dict()
+    with open(analogies, "r") as analogy_f:
+        key = None
+        for line in analogy_f:
+            if line in ['\n', '\r\n']:
+                continue
+            if line.startswith(":"):
+                # new section starts
+                key = line.rstrip()
+                analogy_dict[key] = list()
+            else:
+                analogy_dict[key].append(line[:-1].split("\t"))
+    return analogy_dict
+
+
+def analogy_evaluation(models_path, analogies, first_n):
+    """
+    Compute performance of the model on an analogy test set.
+    :param models_path: string
+    :param analogies: dict
+    :param first_n: int
+    :return: DataFrame
+    """
+    result = pd.DataFrame(
+        columns=["Model", "Type of questions", "Accuracy1 (for present questions)", "Accuracy2 (for all questions)",
+                 "Number of questions", "Number of present questions", "No words in vocabulary"])
+
+    for file in listdir(models_path):
+        if not isfile(join(models_path, file)):
+            continue
+        name = file.replace(".", "_")
+        logger.info(f"Model {name} loading...")
+
+        # TODO: Ideally here we should check the file extension and load binary fb vectors or textual w2v
+        model = gensim.models.fasttext.load_facebook_vectors(models_path + file)
+
+        logger.info("-------------------------------------------------------")
+        logger.info("Starting evaluation")
+
+        total_analogies = len(list((itertools.chain(*analogies.values()))))
+
+        total_correct = 0
+        present_analogies = 0
+
+        for key, questions in analogies.items():
+            section_correct = 0
+            total_section = len(questions)
+            present_analogies += total_section
+            # TODO: validate how many tasks for each category do we have?
+            for i in range(len(questions)):
+                try:
+                    sim = list(list(zip(*(
+                        model.most_similar(positive=[questions[i][0], questions[i][3]], negative=[questions[i][2]],
+                                           topn=first_n))))[0])
+                    if (questions[i][1] in sim
+                            or questions[i][1].lower() in sim
+                            or questions[i][1] in [s.lower() for s in sim]
```

Comment:
that is a bit inefficient and over-engineered.
I think it should be just one clause `questions[i][1].lower() in [s.lower() for s in sim])` instead of four.